File: server/src/NetworkService/NetNamespace.java
package NetworkService;

import Signal.*;
import org.java_websocket.WebSocket;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.UUID;


public class NetNamespace {
    private String name;
    private NetworkService network;
    public Signal entries;
    NetNamespace(String name, NetworkService network){
        this.name = name;
        this.network = network;
        entries = new Signal();
    }
    public Connection connect(NetworkCallBack callBack){
        return entries.connect(new Event<Tuple<WebSocket, JSONObject>>() {
            @Override
            public void run(Tuple<WebSocket, JSONObject> info) throws Exception {
                System.out.println(info.obj1);
                WebSocket user = info.obj1;
                String message = info.obj2.getString("message");

                System.out.println(info.obj1);
                sendBack(user, callBack.run(user, message), info.obj2.getString("callId"));
            }
        });
    }
    public void send(WebSocket user, String message) throws JSONException {
        this.sendClient(user, message, "none");
    }

    public void sendBack(WebSocket user, String message, String callId) throws JSONException {
        this.sendClient(user, message, callId);
    }

    private void sendClient(WebSocket user, String message, String callId) throws JSONException{
        JSONObject json = new JSONObject();
        json.put("name", this.name);
        json.put("messageId", this.generateUUID());
        json.put("callId", callId);

        float limitLength = 30000;
        double nbrSplit = Math.ceil(message.length()/limitLength);

        for(int i = 0; i < nbrSplit; ++i) {
            String pack = message.substring(i*(int)limitLength, Math.min((i+1) * (int)limitLength, message.length()));
            json.put("ended", i+1 == nbrSplit);
            json.put("message", pack);

            user.send(json.toString());
        }
    }
    private String generateUUID() {
        return UUID.randomUUID().toString();
    };
}

File: client/js/NetworkController/NetNameSpace.js
export class NetNameSpace {
    name;
    connectedFunctions = {};
    calls = {};
    network;
    constructor(name, network) {
        this.name = name;
        this.network = network;
    }

    _fireConnections(message){
        for (const key in this.connectedFunctions) {
            this.connectedFunctions[key](message);
        }
    }

    resumeCall(callId, message, error){
        const call = this.calls[callId]
        if(call == null) console.error("call expired !");
        this.calls[callId] = null;

        if(error){
            call.reject(message);
        }else{
            call.resolve(message);
        }
    }

    send(message){
        this._send(message, "none")
    }

    call(message){
        const callId = this._generateUUID()
        let calls = this.calls;
        return new Promise((resolve, reject) => {;
            this.calls[callId] = {
                resolve: resolve,
                reject: reject
            };
            setTimeout(reject, 30000);
            this._send(message, callId);
        }).catch(function (message){
            console.error(message);
            if(calls[callId]) calls[callId] = null;
        });
    }
    _send(message, callId){
        const messageId = this._generateUUID();
        const limitLength = 30000;
        const nbrSplit = Math.ceil(message.length/limitLength);
        console.log(message.length);
        for(let i = 0; i < nbrSplit; ++i){
            const pack = message.substring(i*limitLength, (i+1)*limitLength);

            this.network._send(JSON.stringify({
                callId: callId,
                messageId: messageId,
                ended: i+1 == nbrSplit,
                name: this.name,
                message: pack,
            }))
        }
    }

    connect(foo){
        let functionId = this._generateUUID();
        this.connectedFunctions[functionId] = foo;
        return functionId;
    }

    disconnect(functionId){
        this.connectedFunctions[functionId] = null;
    }
    _generateUUID() { // Public Domain/MIT
        let d = new Date().getTime();//Timestamp
        let d2 = ((typeof performance !== 'undefined') && performance.now && (performance.now()*1000)) || 0;//Time in microseconds since page-load or 0 if unsupported
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
            let r = Math.random() * 16;
            if(d > 0){
                r = (d + r)%16 | 0;
                d = Math.floor(d/16);
            } else {
                r = (d2 + r)%16 | 0;
                d2 = Math.floor(d2/16);
            }
            return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
        });
    }
}